src/dft/backends/cufft/commit.cpp

```diff
@@ -323,6 +351,76 @@ class cufft_commit final : public dft::detail::commit_impl<prec, dom> {
+        }
+    }
+
+    void set_buffer_workspace(cufftHandle plan, sycl::buffer<scalar_type>& bufferWorkspace,
+                              std::size_t range, std::size_t offset) {
+        this->get_queue().submit([&](sycl::handler& cgh) {
+            auto workspace_acc =
+                bufferWorkspace.template get_access<sycl::access::mode::read_write>(cgh, range,
+                                                                                    offset);
+            cgh.host_task([=](sycl::interop_handle ih) {
+                auto stream = ih.get_native_queue<sycl::backend::ext_oneapi_cuda>();
+                auto result = cufftSetStream(plan, stream);
+                if (result != CUFFT_SUCCESS) {
+                    throw oneapi::mkl::exception(
+                        "dft/backends/cufft", "set_workspace",
+                        "cufftSetStream returned " + std::to_string(result));
+                }
+                auto workspace_native = reinterpret_cast<scalar_type*>(
+                    ih.get_native_mem<sycl::backend::ext_oneapi_cuda>(workspace_acc));
+                cufftSetWorkArea(plan, workspace_native);
+            });
+        });
+        this->get_queue().wait_and_throw();
+    }
+
+    virtual void set_workspace(sycl::buffer<scalar_type>& bufferWorkspace) override {
+        this->external_workspace_helper_.set_workspace_throw(*this, bufferWorkspace);
+        std::size_t plan0_count =
+            static_cast<std::size_t>(get_plan0_workspace_size_bytes()) / sizeof(scalar_type);
+        std::size_t total_workspace_count =
+            static_cast<std::size_t>(this->get_workspace_external_bytes()) / sizeof(scalar_type);
+        std::size_t plan1_count = total_workspace_count - plan0_count;
+        buffer_workspace = bufferWorkspace;
+        if (plans[0]) {
+            set_buffer_workspace(*plans[0], bufferWorkspace, plan0_count, 0);
+        }
+        if (plans[1]) {
+            set_buffer_workspace(*plans[1], bufferWorkspace, plan1_count, plan0_count);
+        }
+    }
+
+    std::int64_t get_plan_workspace_size_bytes(cufftHandle handle) {
+        std::size_t size = 0;
+        cufftGetSize(*plans[0], &size);
+        std::int64_t padded_size = static_cast<int64_t>(size);
+        return padded_size;
+    }
+
+    // Separate workspaces are needed for the forward and backward plans. Because the plans[1]'s workspace is offset
+    // by the size of plans[0]'s workspace, we need to be able to get plans[0]'s size individually.
+    std::int64_t get_plan0_workspace_size_bytes() {
+        if (!plans[0]) {
+            return 0;
+        }
+        return get_plan_workspace_size_bytes(*plans[0]);
+    }
+
+    virtual std::int64_t get_workspace_external_bytes_impl() override {
+        std::int64_t size = plans[1] ? get_plan_workspace_size_bytes(*plans[1]) : 0;
+        return size + get_plan0_workspace_size_bytes();
```

Comment:
I didn't think of the issue of multiple plans with the external workspace. I think we should return the maximum of the 2 workspace size. I'm concerned that this would artificially increase the workspace size for users that only use one direction.
This would add the requirement that we must wait between `compute` functions of a same descriptor that uses an external workspace which I think is acceptable.

Reply:
I'd prefer not to add a user requirement about waits between compute calls. But I think we can avoid this:
* I think the buffer implementation already will already have implicit dependencies between compute forward / compute backwards, since they're getting access to the same underlying buffer. 
  * I kept it simple doing this. Anyone migrating from CUDA probably isn't using a buffer anyway.
* For USM we can always have a dependency on the last use of the allocation.
  * I don't *think* it can cause hidden deadlocks. Open to thoughts on that though.

Finally this might be good because in the implementation of this PR I introduced errors in both rocFFT and cuFFT when splitting up the buffers/USM allocs. 

Are you happy with the idea of implicit dependencies in the USM impl?

Reply:
I think the synchronisation should be offloaded to the user. In much the same way that the user shouldn't try to launch two kernels that use the same USM memory without the use of events to synchronize, then it should be clear from our specification that the external workspace is in use while the `compute_forward` kernel is in flight.
Kernels launched from the same descriptor should be independent.
Trying to sort this ourselves goes against USM I would say.

Reply:
I've implemented this with implicit USM dependencies in [9c2b2ff](https://github.com/oneapi-src/oneMKL/pull/411/commits/9c2b2ffc008faf9b5da5a137fa53e2e74f935785) and [19869bf](https://github.com/oneapi-src/oneMKL/pull/411/commits/19869bf06518db4983183a24ca7f68f907f1c834).

I agree that this doesn't fit with the typical usage of USM. However, anyone porting from closed-source *oneMKL* and expecting their code to work with *cuFFT* and *rocFFT* would expect their code to just work without additional changes to dependency tracking.

I'm also under the impression that CUDA streams are in-order (perhaps I'm wrong here - I've never used CUDA directly). Consequently, kernels launched from the same sycl queue would not run concurrently anyway. *cuFFT* launches multiple kernels per DFT, so the USM dependency would enforce that kernels from different DFTs could not be interleaved. In this context, the implicit dependency for the USM implementation does not seem so bad.

Reply:
I'm having trouble finding the docs about closed-source oneMKL synchronizing on the use of the workspace, do you have a link?

CUDA streams are in-order, but we change the stream for each call to `compute_xxxward` allows transforms to happen concurrently.

I think launching N kernels from the same descriptor is similar to launching 1 kernel from each of N descriptors that share the same external workspace. If you had limited memory then you could save a lot by sharing the same workspace between all the descriptors.

Reply:
The reference for closed-source oneMKL is [here](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2023-0/user-allocated-workspaces.html). It states that:
> The descriptor assumes that the workspace will be solely used for FFT, so the user must be careful not to use the same workspace area for other purposes in parallel or before finishing compute_forward/compute_backward (especially for USM as explicit synchronization is required).

My interpretation of this is that for anything other than FFTs, explicit synchronisation is required. Therefore we don't have to worry about synchronisation for multiple DFTs with a single descriptor. Perhaps @raphael-egan can comment?

Regarding CUDA streams, my understanding of how streams map to CUDA was wrong. Thanks for clearing that up.

Reply:
Thanks for the link!
To me, that reads only as a reminder that the descriptor cannot detect if the memory is being used elsewhere. I don't see a guarantee of synchronization between calls to compute_xxxward.

Reply:
It looks as if a number of backends don't make any gaurantees regarding calling multiple *compute_XXX* functions. For example, I believe that calling multiple *cuFFT* compute functions without interdependencies using the same descriptor could lead to problems in the current oneMKL implementation since the SYCL runtime will schedule those without explicit dependencies to run concurrently by changing streams - see [how cuFFT doesn't suppor this on Stackoverflow](https://stackoverflow.com/questions/39030949/cufftsetstream-causes-garbage-output-am-i-doing-something-wrong).

The problem of not scheduling concurrent `compute_*` calls therefore has a larger scope than just for explict workspace implementation, and I believe this should be addressed at a oneAPI spec level.

I'll be removing the interdepencies from the USM version of the code. For SYCL spec conformance, using accessors on workspace buffers is still the correct thing to do, so these will still implicitly result in implicit synchronisation.

Reply:
That looks good to me so I'll resolve this big discussion, thanks.